### PR TITLE
v0.19 upgrade

### PR DESCRIPTION
**Pull request by dessalines**
@MV-GH this is ready for you to look at.

To summarize my changes:

- Upgraded to the new auth method.
- Upgraded all the structs, and fixed some issues with the `copy_generated_types` script.
- Added the new pagination_cursor method.
- Ran a code analysis and fixed a lot of issues there.
- Removed all the version-checking sorting enums. 

I tested most things, and they seem to be working correctly.

Limitations:

- No backwards-compatibility. 
  - I didn't change the MINIMUM_VERSION in the code, because viewing the site does work.
- The `validate_auth` sections of the code: I'm not entirely sure how they should work, but that new endpoint is deployed on voyager.lemmy.ml .

Next week we'll be testing some `0.19` on lemmy.ml to catch any last bugs, which will unfortunately probably break apps for a few days, before we do the `0.19.0` lemmy backend release.

I defer you to you on the best way to handle the upgrade to your new backwards-compatibility library / method. Even if we don't get it merged in time for this release, it'll still be great to have for future ones. I hope we don't get trapped in merge hell with our two PRs, but its not the end of the world if we do, we'll work through it. LMK anything I can do to help there.

cc @nutomic

### CODE REVIEW

**Comment by MV-GH** [comment]
My Api is getting pretty far too. Doing initial integration attempts. Not sure what the timeline is for v0.19. I probably could get it done in 2weeks. Really depends in what problems in run into.

btw this has a good guide for publishing to maven

https://central.sonatype.org/publish/publish-maven/

**Comment by dessalines** [comment]
Cool. Its tough to say yet when we'll release `v0.19`, definitely not this week, but possibly next week. Obvi I'll make sure jerboa is ready before then. But for now at least I'm just planning on supporting v0.19, considering all the API changes.

If you get your backwards-compatible API done soonish tho, we can use that one instead / in favor.

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (MV-GH) [inline app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: This has some other consequences, there was a reason I put all the way at the end. If the site call fails it will still not allow you to login and force you to retry which is good. For example trying to login with 0.19 now will fail at getSite call but it wont cause you to be logged in. Which would be the case now. The check version check would be even more useless as it would set the client regardless

File: app/src/main/java/com/jerboa/model/SiteViewModel.kt
Comment (MV-GH) [inline app/src/main/java/com/jerboa/model/SiteViewModel.kt:51]: Don't forget that you need to set a new auth when you change ur account/log out. Look in AccountViewModel (take over the design, I did with my pr) Encapsulate it so we dont have to worry with removeAccount and etc

**Review by MV-GH — CHANGES_REQUESTED** [review]
I was originally gonna just take over the changes from this PR into mine. But that no longer seems manageable, rebase it will be then not easy for sure. 

I don't agree with some of the analysis changes

For the verification changes, the validateAuth, , success -> good, 4XX -> bad, 500 -> retry. So if validateAuth has the same behaviour is good. I have emulate validateAuth with that getMentionsCall for my API (for 0.18). 


Also might have missed it but there should also be changes to the settings, the removal of that one. The postnotif setting do we completely remove it our just the UI parts. (Idc tbh)

**Comment by MV-GH** [comment]
Also you will have to redo the generate Types thing, there were some changes again

File: app/src/main/java/com/jerboa/api/Http.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/api/Http.kt:323]: I'll get rid of this altogether, since auth is already passed through that interceptor, and the lemmy back-end accepts it now.

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: Not sure I follow this one. Right now it is:

- Create a temp API for logging in.
- If the login succeeds, change the API in order to set the JWT.
- Fetch the site to get `MyUserInfo`
- Save it to the DB
- Go home

What am I missing?

File: app/src/main/java/com/jerboa/model/SiteViewModel.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/model/SiteViewModel.kt:51]: The `changeLemmyInstance` on L54 sets the auth.

File: app/src/main/java/com/jerboa/JerboaAppState.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/JerboaAppState.kt:183]: Fixed.

File: app/src/main/java/com/jerboa/datatypes/Others.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/datatypes/Others.kt:20]: Thx, got it.

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (MV-GH) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: Fetch Site can fail for multiple reasons.

Currently it fails for incompatible versions.  login succeeds for 0.19 but getSite doesn't in this case you do not want to confirm the api. Currently (with your code) it will display error, login failed but it actually successfully changed the api but it didn't make account. Meaning when ppl back out they are using the new api(for the new account) instead of the actual account that is logged in.

**Comment by dessalines** [comment]
K, addressed some of these above. Re-running `copy_generated_types` didn't seem to pick up anything new.

File: app/src/main/java/com/jerboa/model/SiteViewModel.kt
Comment (MV-GH) [reply app/src/main/java/com/jerboa/model/SiteViewModel.kt:51]: Ah yes this is a flow updates the api. mb

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: Hrm. Shouldn't a failed login be caught in the code just above this?

Comment (dessalines) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: Maybe I should be doing the checks on `tempApi.getSite`, and if those pass, then change the instance and continue with the DB stuff?

Comment (MV-GH) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: > Or maybe, login is succeeding, but GetSite fails with older versions?

Yes, not just with older versions also newer versions.

```kt
       API.changeLemmyInstance(instance, jwt)

            // Fetch the site to get your name and id
            siteViewModel.siteRes = apiWrapper(API.getInstance().getSite())

            try {
                when (val siteRes = siteViewModel.siteRes) {
                    is ApiState.Failure -> {
                        val txt = siteRes.msg.message ?: "FAILURE: NO MESSAGE, probably that version not supported"
                        Toast.makeText(ctx, txt, Toast.LENGTH_SHORT).show()
                        throw RuntimeException(txt)
                    }
```

Here we check if the getSite call failed. and show a error msg


The flow when logging with older version:
- Create a temp API for logging in.
- If the login succeeds, change the API in order to set the JWT.
- Confirms the API instance to the new instance
- Fetch the site to get MyUserInfo
- Site Call fails
- Throw error, not supported version, show this
- User backs out the login
- Screen throws errors everywhere, cuz its doing calls with the newly attempted user to login


Before
- Create a temp API for logging in.
- If the login succeeds, change the API in order to set the JWT.
- Fetch the site to get MyUserInfo
- Site Call fails
- Throw error, not supported version, show this
- User backs out the login
- Everything still works because it keeps using the old instance

So just keep the old logic and confirm the api as last as before

Comment (MV-GH) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: > Maybe I should be doing the checks on tempApi.getSite, and if those pass, then change the instance and continue with the DB stuff?

We already do this

File: app/src/main/java/com/jerboa/model/CommunityViewModel.kt
Comment (MV-GH) [inline app/src/main/java/com/jerboa/model/CommunityViewModel.kt:72]: How does this resetPage logic work? in Appending it resets the value regardless the last pageCursor. Does this reset and cursor state even have purpose? Would the reset not be better be to remove the oldRes itself?

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt:63]: Here's the issue on L63-64 : 

```kt
// Fetch the site to get your name and id
siteViewModel.siteRes = apiWrapper(api.getSite(GetSite(auth)....
```

Its using the tempApi, with a custom JWT, to fetch `MyUserInfo` . That's no longer possible.

I'll see if I can refactor it in a way that works, where it only changes it at the end.

File: app/src/main/java/com/jerboa/model/CommunityViewModel.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/model/CommunityViewModel.kt:72]: If pageCursor is null / missing, then it just fetches the first page. If you call `getPosts(page_cursor = ...`, then it fetches the next page of results.

If you trace down resetPage(), you'll see its called for things like changing the sort type, clicking refresh, etc.

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (dessalines) [reply app/src/main/java/com/jerboa/model/LoginViewModel.kt]: Okay I think I got this one now, check the most recent commit or a diff of this file from main.

**Comment by dessalines** [comment]
BTW I also tested image uploading on voyager.lemmy.ml from this PR (I removed that cookie stuff), and it works.

File: app/src/main/java/com/jerboa/model/LoginViewModel.kt
Comment (MV-GH) [inline app/src/main/java/com/jerboa/model/LoginViewModel.kt:60]: small typo

**Comment by MV-GH** [comment]
I was trying to test this PR but voyager seems to be having issues atm. I will test it later

**Comment by MV-GH** [comment]
Alright it works but it seems constantly display, this account is being verified on return out of a post. (It should not even be verifying the account)

Using 0.18 accs seems to work better than I expected, I can browse but it does display the verification thingy stuck on jwt (prob cuz that doesn't fall back thus jwt verification fails as auth=null) (Maybe a warning? we can keep as is but we will definitely get issues around this)

**Review by MV-GH — APPROVED** [review]
I can't reproduce the previous "The account is being verified" issue anymore but there is definitely an edgecase somewhere where it doesn't properly purge the lock

**Comment by dessalines** [comment]
> Using 0.18 accs seems to work better than I expected, I can browse but it does display the verification thingy stuck on jwt (prob cuz that doesn't fall back thus jwt verification fails as auth=null) (Maybe add a version warning? we can keep as is but we will definitely get issues around this)

Where would be the best place to add that toast?

**Comment by MV-GH** [comment]
We had one in root our app by twiz, but that one is based of getSite which wouldn't work as getSite doesn't work across version. You could reuse it but add that nodeinfo stuff to get version but that's a lotta work. Or you could wait until my api is done.(merge as is, release in the meantime) Or we can use my LemmyApi.getVersion that does functionality but we would have cut a prelease first.

**Comment by dessalines** [comment]
I'll wait until your API is done, I'll just merge this as is. Again, sry about the conflicts, lmk if there's anything I can do to help there.

**Comment by dessalines** [comment]
(3) Makes the most sense to me also. Could be a part of the `changeInstance` function, which I believe is called on startup, as well as when switching accounts.

**Pull request merged**
